packages/flutter_test/lib/src/finders.dart

```diff
@@ -460,57 +469,544 @@ class CommonFinders {
+  ///
+  /// {@template flutter_test.finders.FinderBase.requirements}
+  /// This string should be a brief English phrase describing the finder's
+  /// requirements for a single candidate. It should make sense as the end of
+  /// the sentence "Found an item <requirements>".
+  ///
+  /// As an example, a `FinderBase<Object>` that searches for a candidate of a
+  /// specific type would provide 'of type MyObject', so that the example
+  /// sentence comes out to: "Found an item of type MyObject."
+  /// {@endtemplate}
+  ///
+  /// See also:
+  ///
+  ///   * [FinderBase.toString] where this is used to fully describe the finder
+  ///   * [FinderResult.toString] where this is used to provide context to the
+  ///     results of a search
+  String get requirements;
+
+  /// Returns all of the items that will be considered by this finder.
+  @protected
+  Iterable<TCandidate> get allCandidates;
+
+  /// Returns a variant of this finder that only matches the first item
+  /// found by this finder.
+  FinderBase<TCandidate> get first => _FirstFinder<TCandidate>(this);
+
+  /// Returns a variant of this finder that only matches the last item
+  /// found by this finder.
+  FinderBase<TCandidate> get last => _LastFinder<TCandidate>(this);
+
+  /// Returns a variant of this finder that only matches the item at the
+  /// given index found by this finder.
+  FinderBase<TCandidate> at(int index) => _IndexFinder<TCandidate>(this, index);
+
+  /// Returns all the items in the given list that match this
+  /// finder's requirements.
+  ///
+  /// When implementing your own finders that inherit directly from
+  /// [FinderBase], this is the main method to override. If your finder
   /// can efficiently be described just in terms of a predicate
-  /// function, consider extending [MatchFinder] instead.
-  Iterable<Element> apply(Iterable<Element> candidates);
+  /// function, consider mixing in [MatchFinderMixin] instead.
+  @protected
+  Iterable<TCandidate> findInCandidates(Iterable<TCandidate> candidates);
 
-  /// Whether this finder skips nodes that are offstage.
+  /// Searches all of the candidates and returns the result of that search.
   ///
-  /// If this is true, then the elements are walked using
+  /// See also:
+  ///
+  ///   * [found] which will return the latest results without re-executing the
+  ///     search.
+  ///   * [tryFind] which will indicate whether any results were found rather
+  ///     than directly returning results.
+  FinderResult<TCandidate> find() {
+    _found = FinderResult<TCandidate>(this.requirements, findInCandidates(allCandidates));
+    return found;
+  }
+
+  /// Searches all of the candidates and returns whether the search found any
+  /// matching candidates.
+  ///
+  /// This is useful in cases where an action needs to be repeated while or
+  /// until a finder has results. The results from the search can be accessed
+  /// using the [found] property without re-executing the search.
+  bool tryFind() {
```

Comment:
It is not immediately clear to me why we have this (and the `found`) API. Once could just call find and check id the result is empty (or use the result instead of accessing `found`).

Reply:
I added a (probably overly verbose) code example showing how to use it. I'll probably pull out most of the UI code in that example, but hopefully it makes the intent more clear?

Reply:
Cleaned up the code example a bit.

Reply:
I guess it is a slight simplification over

```dart
  var result;
  while ((result = loadedFinder.find()).isEmpty) {
    await tester.pump(const Duration(milliseconds: 100));
  }
  expect(result, hasLength(1));
```

I am still not super convinced that this needs special API, but I am also not super against it.

Reply:
Resolving per discussion over chat. The `tryFind` will be renamed to `tryEvaluate` to remain consistent with `evaluate`, and the `tryEvaluate` method will cover one of the common use cases of `precache` more explicitly. The `precache` method will be deprecated and replaced by `tryEvaluate` and `runCached` depending on the specific usage of the method.